src/l1_fetcher.rs

```diff
@@ -0,0 +1,123 @@
+use ethers::{abi::Contract, prelude::*, providers::Provider};
+use eyre::Result;
+use state_reconstruct::parse_calldata;
+use state_reconstruct::CommitBlockInfoV1;
+use state_reconstruct::BLOCK_STEP;
+use state_reconstruct::GENESIS_BLOCK;
+use state_reconstruct::ZK_SYNC_ADDR;
+use std::sync::Arc;
+use tokio::sync::mpsc;
+
+pub struct L1Fetcher {
+    provider: Provider<Http>,
+    contract: Contract,
+}
+
+impl L1Fetcher {
+    pub fn new(http_url: &str) -> Result<Self> {
+        let provider =
+            Provider::<Http>::try_from(http_url).expect("could not instantiate HTTP Provider");
+
+        let abi_file = std::fs::File::open("./IZkSync.json")?;
+        let contract = Contract::load(abi_file)?;
+
+        Ok(L1Fetcher { provider, contract })
+    }
+
+    pub async fn fetch(
+        &self,
+        sink: mpsc::Sender<Arc<Vec<CommitBlockInfoV1>>>,
+        start_block: Option<U64>,
+        end_block: Option<U64>,
+    ) -> Result<()> {
+        let mut current_l1_block_number = start_block.unwrap_or(GENESIS_BLOCK.into());
+        let end_block_number = end_block.unwrap_or(
+            self.provider
+                .get_block(BlockNumber::Latest)
+                .await?
+                .unwrap()
+                .number
+                .unwrap(),
+        );
+
+        let event = self.contract.events_by_name("BlockCommit")?[0].clone();
+        let function = self.contract.functions_by_name("commitBlocks")?[0].clone();
+
+        let (hash_tx, mut hash_rx) = mpsc::channel(5);
+        let (calldata_tx, mut calldata_rx) = mpsc::channel(5);
+
+        // Split L1 block processing into three async steps:
+        // - BlockCommit event filter.
+        // - Referred L1 block fetch.
+        // - Calldata parsing.
+        let provider = self.provider.clone();
+        tokio::spawn(async move {
+            while let Some(hash) = hash_rx.recv().await {
+                let tx = match provider.get_transaction(hash).await {
+                    Ok(x) => x,
+                    Err(e) => {
+                        println!("failed to get transaction for hash {}: {}", hash, e);
+                        continue;
+                    }
+                };
+
+                let tx = match tx {
+                    Some(tx) => Arc::new(tx.input.to_owned()),
```

Comment:
```suggestion
                    Some(tx) => Arc::new(tx.input.clone()),
```
Nit: To be more explicit, I prefer reserving `to_owned` for when converting from one type to another.

Reply:
What's the difference here? I specifically don't want to `clone()`. This was the place I was trying to figure out how to move the ownership _without_ copying the data. This `tx.input` is the calldata byte array and ideally I'd like to avoid copying that when there's no need for that.

Reply:
Alright, so `to_owned` will *convert* the type and turn it into its owned equivalent (many types this is equivalent to a `clone`). For example, a `&str` will become a `String` which is a different type than `&str`. However, `clone` will instead return the identical type, for our example, `&str`. In cases like this then, I think it's clearer to use `clone()`, if we're actually not changing the underlying type.

Reply:
Ack :+1: